includes/class-wc-subscription.php

```diff
@@ -1913,12 +1913,13 @@ public function payment_failed( $new_status = 'on-hold' ) {
 		$this->add_order_note( __( 'Payment failed.', 'woocommerce-subscriptions' ) );
 
 		// Allow a short circuit for plugins & payment gateways to force max failed payments exceeded
-		if ( 'cancelled' == $new_status || apply_filters( 'woocommerce_subscription_max_failed_payments_exceeded', false, $this ) ) {
-			if ( $this->can_be_updated_to( 'cancelled' ) ) {
-				$this->update_status( 'cancelled', __( 'Subscription Cancelled: maximum number of failed payments reached.', 'woocommerce-subscriptions' ) );
-			}
-		} elseif ( $this->can_be_updated_to( $new_status ) ) {
-			$this->update_status( $new_status );
+		$status_note = '';
+		if ( apply_filters( 'woocommerce_subscription_max_failed_payments_exceeded', false, $this ) ) {
+			$status_note = __( 'Subscription Cancelled: maximum number of failed payments reached.', 'woocommerce-subscriptions' );
```

Comment:
We need to also set `$new_status = 'cancelled';` when this filter is used to preserve backwards compatibility.

Reply:
Gotcha! I have updated the whole implementation again to make this small change easier to read in https://github.com/Automattic/woocommerce-subscriptions-core/pull/687/commits/b7a92f35657d25a5480fdf8e4f15c753506ffd22